### ISSUE TITLE
[explicit-module-boundary-types] Complains about higher order functions that are not directly typed

### ISSUE THREAD

**Issue by essenmitsosse** [typescript-eslint/typescript-eslint#2173]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": [2, {"allowExpressions": true}],
    "@typescript-eslint/explicit-module-boundary-types": [2]
  }
}
```
```TS
export function foo(): (n: number) => (m: number) => string {
    return function(n) {
            return function(m) {  return String(n+m) }
    }
}

export const foo = 
    (): (n: number) => (m: number) => string => 
        (n) => 
            (m) => String(n+m)

export const bar: () => (n: number) => string = 
    () => 
        (n) => String(n)

type Buz = () => (n: number) => string 

export const buz: Buz = 
    () => 
        (n) => String(n)
```

**Expected Result**

Shouldn't give an error for any function where the return type is in any way explicitly typed.

**Actual Result**

Throws errors, for anything that is more then one function deep into a higher order return type. So either higher order functions 3-levels deep or higher order functions, where the whole function is typed instead of just the return type of the top level higher order function.
```
Missing return type on function. eslint(@typescript-eslint/explicit-module-boundary-types)
Argument 'm' should be typed. eslint(@typescript-eslint/explicit-module-boundary-types)

Missing return type on function. eslint(@typescript-eslint/explicit-module-boundary-types)
Argument 'n' should be typed. eslint(@typescript-eslint/explicit-module-boundary-types)
```

**Additional Info**

Reproduction repo: https://github.com/essenmitsosse/typescript-eslint-explicit-module-boundary-types

Thanks by the way for so quickly fixing the first issue.

Issue also appears in 3.1.1-alpha.4. So it isn't fixed by #2169

**Versions**

package | version
-- | --
@typescript-eslint/eslint-plugin | 3.1.0
@typescript-eslint/parser | 3.1.0
TypeScript | 3.9.3
ESLint | 7.1.0
node | 13.14.0
npm | 6.14.4
